### PR TITLE
Add more fields to contacts

### LINKED ISSUE #164: Add more fields to Contacts
Make it possible to access any field that contacts exposes on iOS/Android, see https://github.com/joshuapinter/react-native-unified-contacts#contact-object for more information.

### PR DESCRIPTION
**Pull request by aalices**
Fixes #164. 
Exposing all fields available in Android and iOS contacts. 
sample output on iOS (requesting all fields besides IMAGE and THUMBNAIL) : 
https://gist.github.com/aalices/17386b88facc63d03c6ece228479e45d

 - On iOS `image` and `thumbnail` are stored as base64 strings, on Android it is an URI to the internal asset.
 - A field inside `address` has changed from `pobox` to `poBox`
 - Fields that are available only on iOS: `contactType`, `previousLastName`, `image` (Android stores only the thumbnail), `socialProfiles`

Linked API PR: https://github.com/expo/expo-sdk/pull/61

**Test Plan**
Like in existing tests: it has to be tested if the constants are respected and data types are valid ( for example if image data is possible to display )

### CODE REVIEW

File: ios/Exponent/Versioned/Modules/Api/EXContacts.m
Comment (brentvatne) [inline ios/Exponent/Versioned/Modules/Api/EXContacts.m]: I can imagine in the future that we would rather pass this as a Blob so we don't have to pass the base64 image over the bridge. Is the image data already resized? How big could this data end up being?

**Comment by brentvatne** [comment]
My only significant concern is with respect to the image data, once that's resolved Jesse (works on the Android client) and Ben (works on the iOS client) can do a review pass :)

File: ios/Exponent/Versioned/Modules/Api/EXContacts.m
Comment (aalices) [inline ios/Exponent/Versioned/Modules/Api/EXContacts.m]: The size of `imageDate` is the actual image size on disk so I think the only limit is the device's disk space. Thumbnails have very various sizes - I haven't noticed any defined relationship between those sizes.
We discussed it with Krzysztof and as long as there is no official support for blobs and no ability to get an iOS image URI we can't do much. One of the solutions could be creating some custom Image component that gets an URI, parses it and fetches the desired resource.

Comment (brentvatne) [inline ios/Exponent/Versioned/Modules/Api/EXContacts.m]: @aalices - that makes sense. maybe for image we can use a different API so people have to explicitly opt-in to it and understand the consequences? it seems unlikely that people will want to get full images for multiple contacts at the same time too.

```
let uri = await Contacts.getImageUriForContactAsync(id);
```

Comment (kmagiera) [inline ios/Exponent/Versioned/Modules/Api/EXContacts.m]: I'd rather keep the API the same for both android and iOS. We should perhaps add an appropriate comment in docs so that ppl know who can happen if they request the image field to be returned in the contacts data. Note that we don't send the image data if it hasn't been requested.

I don't see any way on how this could be implemented on iOS without passing the image data over the bridge. The issue is that the data is returned from the iOS API as a part of the contacts record in NSData. There is no way currently for us to retain that data on the native side as even if we do we won't be able to tell when the data is no longer being referenced on the JS side and when it should be released.

One idea I have at the moment would be to introduce a native component "ContactImage" that would take ID of the contact as a parameter. It could extend the regular "Image" component but instead of loading image data from a given source it would use contacts API to get the image data and to get it displayed

Comment (brentvatne) [inline ios/Exponent/Versioned/Modules/Api/EXContacts.m]: @kmagiera - would `let uri = await Contacts.getImageUriForContactAsync(id);` not be possible on both platforms?

Regardless, your `ContactImage` solution sounds reasonable to me on second thought. `Expo.Contacts.Image` could be a good way to namespace that.

**Comment by brentvatne** [comment]
@aalices - can you verify that this PR resolves #200?

File: ios/Exponent/Versioned/Modules/Api/EXContacts.m
Comment (kmagiera) [inline ios/Exponent/Versioned/Modules/Api/EXContacts.m]: @brentvatne together with @aalices we decided we're going to add a method to the JS contacts API that would allow you to fetch contacts data for a single contact. This way you could do a superset of things you can with `getImageUriForContactAsync` (imagine you want to implement contacts app where you list contacts by name on the main view and then when you navigate into details you see all the data, currently contacts API only have one method to fetch all contacts).

On top of that we are going to log a warning (yellow box) whenever you try to fetch image data for a list of contacts and suggest you should only do that for a single contact.

@aalices will be updating this PR soon

**Comment by aalices** [comment]
@brentvatne Turns out that `CommonDataKinds.Phone.NORMALIZED_NUMBER` is `null` in some cases, like when phone number is `(555) 555-5555` - with all the brackets and dashes. Anyway, `CommonDataKinds.Phone.NUMBER` is okay and displays the number as it is typed in the contact, I guess. The former is a E.164 representation of the latter. So I think the only solution for this is using the `NUMBER`

**Comment by brentvatne** [comment]
using `CommonDataKinds.Phone.NUMBER` sounds reasonable to me -- people can format the number however they like in JS :) thanks!

File: ios/Exponent/Versioned/Modules/Api/EXContacts.m
Comment (terribleben) [inline ios/Exponent/Versioned/Modules/Api/EXContacts.m]: Can we call this something else? Having a variable called `id` is confusing in objective-c

File: ios/Exponent/Versioned/Modules/Api/EXContacts.m
Comment (terribleben) [inline ios/Exponent/Versioned/Modules/Api/EXContacts.m]: You can just box this, `@(person.imageDataAvailable)`

File: ios/Exponent/Versioned/Modules/Api/EXContacts.m
Comment (terribleben) [inline ios/Exponent/Versioned/Modules/Api/EXContacts.m]: Should we wrap this in a check as to whether `imageDataAvailable` is flagged true?

File: ios/Exponent/Versioned/Modules/Api/EXContacts.m
Comment (terribleben) [inline ios/Exponent/Versioned/Modules/Api/EXContacts.m]: Same here, are we confident `person.thumbnailImageData` actually exists?

File: ios/Exponent/Versioned/Modules/Api/EXContacts.m
Comment (terribleben) [inline ios/Exponent/Versioned/Modules/Api/EXContacts.m]: Why mutable? Can just return a literal `@{ @"day": ..., @"month": ... }` or `nil`

File: ios/Exponent/Versioned/Modules/Api/EXContacts.m
Comment (terribleben) [inline ios/Exponent/Versioned/Modules/Api/EXContacts.m]: You specify `_Nonnull` in the method footprint, but you check again here, not sure which is intended

File: ios/Exponent/Versioned/Modules/Api/EXContacts.m
Comment (terribleben) [inline ios/Exponent/Versioned/Modules/Api/EXContacts.m]: Can you update this comment as well?

**Review by terribleben — CHANGES_REQUESTED** [review]
Thanks! Looking mostly there on the iOS side, a few questions from me

File: ios/Exponent/Versioned/Modules/Api/EXContacts.m
Comment (aalices) [inline ios/Exponent/Versioned/Modules/Api/EXContacts.m]: I would rather leave it to user to check the availability flag as they might want to fetch the data anyway and get a `{ uri: null }` response. What do you think?

**Comment by aalices** [comment]
Look like there are no more comments to the iOS bit. @brentvatne is there anyone who could review android part?

**Comment by jesseruder** [comment]
android side looks good, but i want to get docs and tests ready before we merge. @brentvatne how do you want to handle this since they're not part of our monorepo?

**Comment by terribleben** [comment]
This is merged to our SDK 18 candidate branch, the PR will close once that lands on master later this week!

**Pull request closed without merging**
